File: backend/app/models/llm_config.py
"""Pydantic request/response schemas for the LLM-config admin API.

Validation rules mirror the design note (Provider taxonomy section).
The conditional 'api_key required for non-ollama providers' is enforced
at the Pydantic boundary so route handlers can assume a valid payload.
"""

from __future__ import annotations

from datetime import datetime
from typing import Annotated, Literal
from uuid import UUID

from pydantic import (
    AnyHttpUrl,
    BaseModel,
    Field,
    StringConstraints,
    field_validator,
    model_validator,
)

ProviderType = Literal[
    "anthropic",
    "openai",
    "google",
    "mistral",
    "openai_compatible",
    "ollama",
]

ModelName = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, max_length=200)]
APIKey = Annotated[str, StringConstraints(min_length=8, max_length=512)]


class LLMConfigCreate(BaseModel):
    """Request body for POST /api/v1/admin/llm-config and POST /test."""

    provider: ProviderType
    model_name: ModelName
    base_url: AnyHttpUrl | None = None
    # `None` or empty means "keep the saved key on update; reject on create
    # for non-ollama providers". The route handler enforces the
    # create-vs-update branch using this field's emptiness.
    api_key: APIKey | None = None

    @field_validator("base_url", mode="before")
    @classmethod
    def _empty_string_is_none(cls, v: object) -> object:
        if isinstance(v, str) and not v.strip():
            return None
        return v

    @model_validator(mode="after")
    def _check_required_combinations(self) -> LLMConfigCreate:
        if self.provider in ("openai_compatible", "ollama") and self.base_url is None:
            raise ValueError(f"base_url is required for provider {self.provider!r}")
        return self


class LLMConfigResponse(BaseModel):
    """Public view of a tenant_llm_configs row.

    The api_key is never returned in full — only `api_key_last4`. The
    `warning` field is non-null when the row was saved without a passing
    test, so the UI can surface it.
    """

    id: UUID
    provider: ProviderType
    model_name: str
    base_url: AnyHttpUrl | None
    api_key_last4: str
    test_status: Literal["untested", "passed", "failed"]
    test_error: str | None
    tested_at: datetime | None
    created_at: datetime
    updated_at: datetime
    warning: str | None = None


class TestResult(BaseModel):
    """Response body for POST /api/v1/admin/llm-config/test."""

    ok: bool
    latency_ms: int = Field(ge=0)
    error: str | None = None

File: frontend/src/lib/api/llm-config.ts
/**
 * Typed fetch wrappers for /api/v1/admin/llm-config.
 *
 * Mirrors the Pydantic schemas in backend/app/models/llm_config.py.
 * V1.5a sends `X-Tenant-ID` as the only auth signal — the real auth
 * layer arrives in V2.
 */

const API_BASE =
  process.env.NEXT_PUBLIC_API_BASE ?? "http://localhost:8000";

export type Provider =
  | "anthropic"
  | "openai"
  | "google"
  | "mistral"
  | "openai_compatible"
  | "ollama";

export type TestStatus = "untested" | "passed" | "failed";

export interface LLMConfigCreate {
  provider: Provider;
  model_name: string;
  base_url?: string | null;
  api_key?: string | null;
}

export interface LLMConfigResponse {
  id: string;
  provider: Provider;
  model_name: string;
  base_url: string | null;
  api_key_last4: string;
  test_status: TestStatus;
  test_error: string | null;
  tested_at: string | null;
  created_at: string;
  updated_at: string;
  warning: string | null;
}

export interface TestResult {
  ok: boolean;
  latency_ms: number;
  error: string | null;
}

export class ApiError extends Error {
  status: number;
  detail: string;
  constructor(status: number, detail: string) {
    super(`${status}: ${detail}`);
    this.status = status;
    this.detail = detail;
  }
}

async function call<T>(
  path: string,
  opts: { method: string; tenantId: string; body?: unknown },
): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    method: opts.method,
    headers: {
      "X-Tenant-ID": opts.tenantId,
      "Content-Type": "application/json",
    },
    body: opts.body === undefined ? undefined : JSON.stringify(opts.body),
    cache: "no-store",
  });
  if (res.status === 204) {
    return null as T;
  }
  const text = await res.text();
  let parsed: unknown = undefined;
  if (text) {
    try {
      parsed = JSON.parse(text);
    } catch {
      /* keep parsed=undefined */
    }
  }
  if (!res.ok) {
    const detail =
      parsed && typeof parsed === "object" && "detail" in parsed
        ? String((parsed as { detail: unknown }).detail)
        : text || res.statusText;
    throw new ApiError(res.status, detail);
  }
  return parsed as T;
}

export const llmConfigApi = {
  get: (tenantId: string): Promise<LLMConfigResponse> =>
    call<LLMConfigResponse>("/api/v1/admin/llm-config", {
      method: "GET",
      tenantId,
    }),

  upsert: (
    tenantId: string,
    body: LLMConfigCreate,
  ): Promise<LLMConfigResponse> =>
    call<LLMConfigResponse>("/api/v1/admin/llm-config", {
      method: "POST",
      tenantId,
      body,
    }),

  delete: (tenantId: string): Promise<null> =>
    call<null>("/api/v1/admin/llm-config", {
      method: "DELETE",
      tenantId,
    }),

  test: (tenantId: string, body: LLMConfigCreate): Promise<TestResult> =>
    call<TestResult>("/api/v1/admin/llm-config/test", {
      method: "POST",
      tenantId,
      body,
    }),
};
